src/sub.c

```diff
@@ -164,18 +182,79 @@ natsSub_deliverMsgs(void *arg)
 void
 natsSub_close(natsSubscription *sub, bool connectionClosed)
 {
+    natsMsgDlvWorker *ldw = NULL;
+
     natsSub_Lock(sub);
 
-    sub->closed = true;
-    sub->connClosed = connectionClosed;
-    natsCondition_Broadcast(sub->cond);
+    SUB_DLV_WORKER_LOCK(sub);
+
+    if (!(sub->closed))
+    {
+        sub->closed = true;
+        sub->connClosed = connectionClosed;
+
+        if (sub->libDlvWorker != NULL)
+        {
+            // If this is a subscription with timeout, stop the timer.
+            if (sub->timeout != 0)
+                natsTimer_Stop(sub->timeoutTimer);
+
+            // Post a control message to wake-up the worker which will
+            // ensure that all pending messages for this subscription
+            // are removed and the subscription will ultimately be
+            // released in the worker thread.
+            natsLib_msgDeliveryPostControlMsg(sub);
+        }
+        else
+            natsCondition_Broadcast(sub->cond);
+    }
+
+    SUB_DLV_WORKER_UNLOCK(sub);
 
     natsSub_Unlock(sub);
 }
 
+static void
+_asyncTimeoutCb(natsTimer *timer, void* closure)
+{
+    natsSubscription *sub = (natsSubscription*) closure;
+
+    // Should not happen, but in case
+    if (sub->libDlvWorker == NULL)
+        return;
+
+    SUB_DLV_WORKER_LOCK(sub);
+
+    // If the subscription is closed, or if we are prevented from posting
+    // a "timeout" control message, do nothing.
+    if (!sub->closed && !sub->timedOut && !sub->timeoutSuspended)
+    {
+        // Prevent from scheduling another control message while we are not
+        // done with previous one.
+        sub->timedOut = true;
+
+        // Set the timer to a very high value, it will be reset from the
+        // worker thread.
+        natsTimer_Reset(sub->timeoutTimer, 60*60*1000);
```

Comment:
That's how my timers work, they will fire again automatically, that is, it is not a one time thing. After creating a timer and assign a callback to the timer, the callback will fire every given interval, without the need to call reset. Here I call reset to "stop" the firing since the worker thread will take care of the reset after the user callback returns.